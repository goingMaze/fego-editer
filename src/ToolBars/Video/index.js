import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { AtomicBlockUtils } from 'draft-js'
import { ToolBarBtn } from 'utils'
import Iconfont from 'iconfont'

const VIDEO_TYPES = [
	{ label: 'video', style: 'video', icon: 'iconVideo', title: '插入视频' }
], defaultState = {
	videoVisible: false,
	urlValue: '',
	btnDisabled: true,
	width: 'auto',
	height: 'auto',
	controls: false,
	autoPlay: true
}

export default class Video extends Component {
	static propTypes = {
		editorState: PropTypes.object.isRequired,
		onChange: PropTypes.func.isRequired,
	}
	constructor(props) {
		super(props)
		this.state = defaultState
	}
	componentWillMount() {
		let { modalManage } = this.props;
		modalManage.addCallback(this.changeVideoVisible)
	}
	componentWillUnmount() {
		let { modalManage } = this.props;
		modalManage.removeCallback(this.changeVideoVisible)
	}
	changeVideoVisible = () => {
		this.setState({
			...defaultState,
			videoVisible: this.videoVisible
		})
		this.videoVisible = false
	}
	confirmMedia = () => {
		const { editorState, onChange } = this.props;
		const { urlValue, urlType, width, height, controls, autoPlay } = this.state;
		const contentState = editorState.getCurrentContent();
		const entityKey = contentState.createEntity(urlType, 'IMMUTABLE', { src: urlValue, width, height, controls, autoPlay }).getLastCreatedEntityKey();
		onChange(AtomicBlockUtils.insertAtomicBlock(editorState, entityKey, ' '))
		this.changeVideoVisible()
	}
	handleVideoAdd = () => {
		this.confirmMedia()
	}
	handleChange = (type, e) => {
		let { value, checked } = e.target
		if (type === 'videoUrl') {
			this.setState({
				urlValue: value,
				btnDisabled: !Boolean(value)
			})
		}
		this.setState({ [type]: e.target.getAttribute('type') !== 'checkbox' ? value : checked })
	}
	onToggle = () => {
		this.videoVisible = !this.state.videoVisible
		this.setState({
			urlType: 'video',
		})
	}
	render() {
		let { videoVisible, urlValue, btnDisabled, width, height, controls, autoPlay } = this.state;
		let videoModal = '';
		if (videoVisible) {
			videoModal = (
				<div className='modal' onMouseDown={e => e.stopPropagation()} >
					<div className='uploadBox' >
						<input type='text' value={urlValue} onChange={this.handleChange.bind(this, 'videoUrl')} placeholder='请输入视频url' className='uploadArea input' />
						<div className='videoStyleDiv' >
							<label>宽度：&nbsp;&nbsp;</label>
							<input type="text" value={width} onChange={this.handleChange.bind(this, 'width')} className='videoStyleInput' />
						</div>
						<div className='videoStyleDiv' >
							<label>高度：&nbsp;&nbsp;</label>
							<input type="text" value={height} onChange={this.handleChange.bind(this, 'height')} className='videoStyleInput' />
						</div>
						<div className='videoStyleDiv' >
							<label>显示控件：&nbsp;&nbsp;</label>
							<input type="checkbox" checked={controls} onChange={this.handleChange.bind(this, 'controls')}  />
						</div>
						<div className='videoStyleDiv' >
							<label>自动播放：&nbsp;&nbsp;</label>
							<input type="checkbox" checked={autoPlay} onChange={this.handleChange.bind(this, 'autoPlay')} />
						</div>
					</div>
					<div className='center' >
						<button className='btn btnPrimary' disabled={btnDisabled} onMouseDown={this.handleVideoAdd} >确定</button> &nbsp;&nbsp;&nbsp;&nbsp;
						<button className='btn' onMouseDown={this.changeVideoVisible} >取消</button>
					</div>
				</div>
			)
		}
		return (
			<div className="FegoEditor-toolbars">
				{
					VIDEO_TYPES.map(type =>
						<ToolBarBtn
							key={type.label} title={type.title}
							onToggle={this.onToggle}
							style={type.style}
						>
							{Iconfont[type.icon]}
						</ToolBarBtn>
					)
				}
				{videoModal}
			</div>
		)
	}
}