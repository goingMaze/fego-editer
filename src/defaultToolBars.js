const defaultToolBars = {
	options: ['BlockType', 'Color', 'FontFamily', 'FontSize', 'Image', 'Inline', 'Link', 'List', 'TextAlign', 'Video', 'History'],
	BgColor: [],
	BlockType: ['正文', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'],
	Color: ['rgb(97,189,109)', 'rgb(26,188,156)', 'rgb(84,172,210)', 'rgb(44,130,201)',
		'rgb(147,101,184)', 'rgb(71,85,119)', 'rgb(204,204,204)', 'rgb(65,168,95)', 'rgb(0,168,133)',
		'rgb(61,142,185)', 'rgb(41,105,176)', 'rgb(85,57,130)', 'rgb(40,50,78)', 'rgb(0,0,0)',
		'rgb(247,218,100)', 'rgb(251,160,38)', 'rgb(235,107,86)', 'rgb(226,80,65)', 'rgb(163,143,132)',
		'rgb(239,239,239)', 'rgb(255,255,255)', 'rgb(250,197,28)', 'rgb(243,121,52)', 'rgb(184,49,47)',
		'rgb(209,72,65)', 'rgb(124,112,107)', 'rgb(209,213,216)'],
	FontFamily: [{
		name: 'SimHei', family: '"SimHei", "黑体"'
	}, {
		name: 'Yahei', family: '"Microsoft Yahei", "微软雅黑"'
	}, {
		name: 'KaiTi', family: '"KaiTi", "楷体"'
	}, {
		name: 'FangSong', family: '"FangSong", "仿宋"'
	}],
	FontSize: [16, 20, 30, 40, 50, 70, 100, 200],
	History: ['undo', 'redo'],
	Image: {
		uploadUrl: '',
		uploadCallback: null
	},
	Inline: ['bold', 'italic', 'underline'],
	Link: {},
	List: ['ul', 'ol', 'indent', 'outdent'],
	TextAlign: ['left', 'center', 'right', 'justify'],
	Video: {},
	Remove: {}
};

export default defaultToolBars
