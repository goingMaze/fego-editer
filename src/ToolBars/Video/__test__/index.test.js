import React from 'react'
import { mount, render } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { EditorState, ContentState, convertFromHTML } from 'draft-js'
import Video from '..'
import defaultToolBars from '../../../defaultToolBars'
import ModalManage from '../../../utils/ModalManage'

describe('Video', () => {
	const contentState = ContentState.createFromBlockArray(convertFromHTML('<div>test</div>'))
	const editorState = EditorState.createWithContent(contentState)
	const wrapper = mount(
		<Video editorState={editorState} onChange={() => { }}
			config={defaultToolBars.Video} modalManage={new ModalManage()}
		/>
	);
	it('渲染时开头应该有一个div', () => {
		expect(wrapper.html().startsWith('<div')).toEqual(true)
	})
	it('should have a toolBarBtn child component', () => {
		expect(wrapper.find('ToolBarBtn').length).toBe(1)
	})
	it('快照', () => {
		const rendered = render(
			<Video editorState={editorState} onChange={() => { }}
				config={defaultToolBars.Video} modalManage={new ModalManage()}
			/>
		)
		expect(toJSON(rendered)).toMatchSnapshot()
	})
})