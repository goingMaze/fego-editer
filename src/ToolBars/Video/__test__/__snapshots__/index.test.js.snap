// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video 快照 1`] = `
<div
  class="FegoEditor-toolbars"
>
  <div
    class="FegoEditor-toolBarBtn "
    title="插入视频"
  >
    <svg
      aria-hidden="true"
      class="icon"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M512 62a450 450 0 1 0 0 900A450 450 0 0 0 512 62z m-1.09285752 835.71428584C282.5 897.71428584 124.87142832 701.38571416 126.28571416 513.02857168c1.35-187.39285752 182.05714248-385.97142833 384.62142832-384.55714336 209.18571416 1.54285752 385.45714248 188.80714248 384.55714336 384.6214292-0.70714248 182.31428584-155.05714248 371.44285752-384.55714336 384.62142832zM383.42857168 319.14285752l321.4285708 192.85714248-321.4285708 192.85714248V319.14285752z"
      />
    </svg>
  </div>
</div>
`;
